indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+      List<SegmentAllocateRequest> requests,
+      AllocateRequestBatch requestBatch
+  )
+  {
+    if (requests.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    final AllocateRequestKey requestKey = requestBatch.key;
+    log.info(
+        "Trying allocation for [%d] requests, interval [%s] in batch [%s]",
+        requests.size(),
+        tryInterval,
+        requestKey
+    );
+
+    final List<SegmentAllocateResult> results = taskLockbox.allocateSegments(
+        requests,
+        requestKey.dataSource,
+        tryInterval,
+        requestKey.skipSegmentLineageCheck,
+        requestKey.lockGranularity
+    );
+
+    final List<SegmentAllocateRequest> successfulRequests = new ArrayList<>();
+    for (int i = 0; i < requests.size(); ++i) {
+      SegmentAllocateRequest request = requests.get(i);
+      SegmentAllocateResult result = results.get(i);
+      if (result.isSuccess()) {
+        successfulRequests.add(request);
+      }
+
+      requestBatch.handleResult(result, request);
+    }
+
+    return successfulRequests;
+  }
+
+  private Map<Interval, List<SegmentAllocateRequest>> getRequestsByInterval(
+      Set<SegmentAllocateRequest> requests,
+      Granularity tryGranularity
+  )
+  {
+    final Map<Interval, List<SegmentAllocateRequest>> tryIntervalToRequests = new HashMap<>();
+    for (SegmentAllocateRequest request : requests) {
+      Interval tryInterval = tryGranularity.bucket(request.getAction().getTimestamp());
+      if (tryInterval.contains(request.getRowInterval())) {
+        tryIntervalToRequests.computeIfAbsent(tryInterval, i -> new ArrayList<>()).add(request);
+      }
+    }
+    return tryIntervalToRequests;
+  }
+
+  private void emitTaskMetric(String metric, long value, SegmentAllocateRequest request)
+  {
+    final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+    IndexTaskUtils.setTaskDimensions(metricBuilder, request.getTask());
+    metricBuilder.setDimension("taskActionType", SegmentAllocateAction.TYPE);
+    emitter.emit(metricBuilder.build(metric, value));
+  }
+
+  private void emitBatchMetric(String metric, long value, AllocateRequestKey key)
+  {
+    final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+    metricBuilder.setDimension("taskActionType", SegmentAllocateAction.TYPE);
+    metricBuilder.setDimension(DruidMetrics.DATASOURCE, key.dataSource);
+    metricBuilder.setDimension(DruidMetrics.INTERVAL, key.preferredAllocationInterval.toString());
+    emitter.emit(metricBuilder.build(metric, value));
+  }
+
+  public void becomeLeader()
+  {
+    if (isEnabled()) {
+      // Start polling the queue
+      log.info("Elected leader. Starting queue processing.");
+      scheduleQueuePoll(maxWaitTimeMillis);
+    } else {
+      log.info("Elected leader but batched segment allocation is disabled.");
+    }
+  }
+
+  public void stopBeingLeader()
+  {
+    log.info("Not leader anymore. Stopping queue processing.");
```

Comment:
should have different logging depending on whether the batch allocation is enabled or not.

Reply:
you could have also added a no-op implementation that throws an exception if used.

Reply:
Do you think a no-op impl would be useful for tests?

Reply:
Hmm, I think having a no-op implementation would require a factory that creates the appropriate `SegmentAllocationQueue` based on the given configs. But that would be cleaner, I suppose.